### DISCUSSION TITLE
New multiplayer mode

### DISCUSSION THREAD

**Discussion by MrStahlfelge** [MrStahlfelge/lightblocks#24] [General]
This discussion can be used to discuss the all-new server multiplayer mode available in beta versions from 1.5.2102 and upwards. Please share your thoughts on the new multiplayer with me here. :)

**Comment by Yonle** [comment]
Hey, So `lightblocks/core/src/de/golfgl/lightblocks/multiplayer/ServerMultiplayerModel.java` is the one who handle the incomming movement in frontend? I think this is need to be improved for high-latency client. Having some server for some regions may still not helps for some cases.

I see these functions is called when movements from player (Which is from server) is received: `handleTetroMoved()`, `handleRotateTetro()`, `handleClearInsert()`, `handleGameOver()`, `handleNextTetro()`, `handleActivateNextTetro()`, `handleSwapHoldAndActive()`, and `handlePinTetro()`.

This is what i think:

- The `handleNextTetro()` should be improved by both clients and server, Atleast 5-10-15 (Depend on user ping) random next tetro should be sent by server (then handled by client) and verified by server (Note that SwapHoldAndActive may need to sent to the server so the server know that user is swapping tetro) to ensure that there's no cheat was used.
- Some player input function like `inputStartMoveHorizontal()` Should be handled instead of only sending movement data.

**Reply by Yonle** [reply]
Seems like `lightblocks/core/src/de/golfgl/lightblocks/model/DeviceMultiplayerModel.java` is the handler in frontend.

**Comment by MrStahlfelge** [comment]
You are right. Feel free to dive in. 

The gameplay as it is is okay as long as the ping to the server is below 50ms. The advantage of having the server take over control completely is that it can handle game modes this way that the client does not know about.

**Reply by Yonle** [reply]
Fact, I'm not a java programmer, But that's my idea as that also applies to other online tetris game.

**Comment by robertino7789** [comment]
On mobile local multiplayer servers there 2 gamemodes - classic and the second with "balance" icon (sorry for my bad English).

It would be good if there were a modern gamemode in local servers, like in Singleplayer practice mode
